### ISSUE TITLE
[support]: tasmanian test failure via spack test

### ISSUE THREAD

**Issue by shahzebsiddiqui** [E4S-Project/e4s#111]
### Name of Software

tasmanian

### Contact Details

shahzebsiddiqui@lbl.gov

### HPC System

Perlmutter

### Request Description

spack is failing to find cmake because it is looking for cmake in the build environment. This is not acceptable since the build environment is removed it was built as part of `spack ci` and $CI_PROJECT_DIR gets removed

```
 ~/ spack test run tasmanian
==> Spack test malx6uzinm7f4jrigzketwym4pbef54j
==> Testing package tasmanian-7.7-4yefyvt
==> Error: TestFailure: 1 tests failed.


Failed to find executable '/global/cfs/cdirs/m3503/ci-data-dir/siddiq90/static/00182220/NERSC/spack-infrastructure/spack/opt/spack/__spack_path_placeholder/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake'

/global/common/software/spackecp/perlmutter/e4s-22.05/78535/spack/lib/spack/spack/package.py:2080, in run_test:
       2077                    # stack instead of from traceback.
       2078                    # The traceback is truncated here, so we can't use it to
       2079                    # traverse the stack.
  >>   2080                    m = '\n'.join(
       2081                        spack.build_environment.get_package_context(tb)
       2082                    )
       2083


/global/common/software/spackecp/perlmutter/e4s-22.05/78535/spack/lib/spack/spack/build_environment.py:1076, in _setup_pkg_and_run:
       1073        tb_string = traceback.format_exc()
       1074
       1075        # build up some context from the offending package so we can
  >>   1076        # show that, too.
       1077        package_context = get_package_context(tb)
       1078
       1079        logfile = None

See test log for details:
  /global/homes/s/siddiq90/.spack/test/malx6uzinm7f4jrigzketwym4pbef54j/tasmanian-7.7-4yefyvt-test-out.txt

======================== 1 failed, 0 passed of 1 specs =========================
==> Error: 1 test(s) in the suite failed.
```

As a normal user i am unable to access the directory so we also have a permission issue

```
 ~/ ls -ld /global/cfs/cdirs/m3503/ci-data-dir/siddiq90/static/00182220/NERSC/spack-infrastructure/spack/opt/spack/__spack_path_placeholder/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake
ls: cannot access '/global/cfs/cdirs/m3503/ci-data-dir/siddiq90/static/00182220/NERSC/spack-infrastructure/spack/opt/spack/__spack_path_placeholder/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake': Permission denied
```

In this event the directory doesn't exist i checked it via `e4s` user

```
(spack-pyenv) e4s:login25> ls -ld /global/cfs/cdirs/m3503/ci-data-dir/siddiq90/static/00182220/NERSC/spack-infrastructure/spack/opt/spack/__spack_path_placeholder/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake
ls: cannot access '/global/cfs/cdirs/m3503/ci-data-dir/siddiq90/static/00182220/NERSC/spack-infrastructure/spack/opt/spack/__spack_path_placeholder/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake': No such file or directory
```

I loaded tasmanian and `cmake` is in the path but it is not using this one for the test

```
 ~/ spack load --list
==> 18 loaded packages
-- cray-sles15-zen3 / gcc@11.2.0 --------------------------------
bzip2@1.0.6  cmake@3.23.1  curl@7.66.0  expat@2.4.8  libarchive@3.5.2  libbsd@0.11.5  libiconv@1.16  libmd@1.0.4  libuv@1.44.1  lz4@1.9.3  lzo@2.10  mbedtls@2.28.0  ncurses@6.1  rhash@1.4.2  tasmanian@7.7  xz@5.2.5  zlib@1.2.12  zstd@1.5.2
 ~/ which cmake
/global/common/software/spackecp/perlmutter/e4s-22.05/78535/spack/opt/spack/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake
```

### Reproduce Bug

```bash
1. `ml e4s/22.05`
2. `spack env activate -V gcc`
3. `spack load tasmanian`
4. `spack test run tasmanian`
```

**Comment by shahzebsiddiqui** [comment]
@mkstoyanov is the maintainer of this package

**Comment by mkstoyanov** [comment]
Tasmanian is a library and a user that wants to link code to the library would naturally need compile time tools, e.g., `cmake` and the compilers, including the hip and cuda compilers. Therefore, the post-install testing for Tasmanian uses `cmake`. The Tasmanian package correctly lists `cmake` as a `run` dependence, I don't understand why that is being removed from the system.

What is the correct way to access `cmake` in a post-install test?

**Comment by wspear** [comment]
If memory serves, spack test does not do *any* package loading. So it is necessary to do `spack load <package> ; spack test run <package>`.

**Comment by mkstoyanov** [comment]
@shahzebsiddiqui shows a log with `spack load --list` that has some version of `cmake` but apparently that is different than the `cmake` used to build Tasmanian. 

@wspear spack documentation does not say that anything gets loaded (as you pointed out); however, run dependencies should be added to the `PATH`.

https://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types

Right now, Tasmanian stores the path at build time, which is a hack that someone at E4S added (not sure who, nor does it matter). How do I get the correct `cmake` path?

**Comment by shahzebsiddiqui** [comment]
yeah this cmake path `/global/cfs/cdirs/m3503/ci-data-dir/siddiq90/static/00182220/NERSC/spack-infrastructure/spack/opt/spack/__spack_path_placeholder/cray-sles15-zen3/gcc-11.2.0/cmake-3.23.1-p23fzuowp4yuitemelic7f65nwybthxd/bin/cmake` is one used by `spack ci generate` + `spack ci rebuild` where each gitlab job does a clone of spack and builds each package. The problem is this test is depending on this build environment which no longer exists. 

The root directory /global/cfs/cdirs/m3503/ci-data-dir/ is part of the gitlab runner root directory and this gets cleaned up.

**Comment by mkstoyanov** [comment]
Why do we clean the dependencies? CMake is one example of a needed dependence, but what if the package links to other libraries? Surely deleting libraries will break the installation of not just Tasmanian but many other packages.

How can we fix this?

**Comment by shahzebsiddiqui** [comment]
This is just how gitlab runner is configured we will run out of quota if we don't do automatic deletion.

Note we can configure the runner to disable deletion but it will cause us to exceed quota very quickly.

I didn't expect spack will look at build directory when installing from buildcache. We basically use the spack ci to populate the buildcache and then do production deployment from cache

**Comment by shahzebsiddiqui** [comment]
This is just how gitlab runner is configured we will run out of quota if we don't do automatic deletion.

Note we can configure the runner to disable deletion but it will cause us to exceed quota very quickly.

I didn't expect spack will look at build directory when installing from buildcache. We basically use the spack ci to populate the buildcache and then do production deployment from cache

**Comment by shahzebsiddiqui** [comment]
> Why do we clean the dependencies? CMake is one example of a needed dependence, but what if the package links to other libraries? Surely deleting libraries will break the installation of not just Tasmanian but many other packages.
> 
> How can we fix this?

I guess if we can get a standalone test for tasmanian that doesn't rely on `spack test` we can use that for testing E4S/22.05. Perhaps we can use E4S Testsuite (@wspear) or you can contribute a test directly into our test repo https://github.com/buildtesters/buildtest-nersc.

Looking at e4s testsuite https://github.com/E4S-Project/testsuite/blob/master/validation_tests/tasmanian/run.sh this test also relies on spack test so E4S Testsuite is probably going to result in same error unless the test gets changed in the Testsuite repo. The other option is to figure out a test that could run on Perlmutter, write up a standalone shell script and then contribute that test to our repo.

If you need any assistance with contribution to  https://github.com/buildtesters/buildtest-nersc, please reach out to me.

Btw, are you in ECP/Spack slack so we can continue conversation on slack

**Comment by wspear** [comment]
Support for spack test issues should be handled on the spack repository tracker. The relevant issue is here: https://github.com/spack/spack/issues/35266
